### PR TITLE
chore: remove call to legacy Charm() on bootstrap

### PR DESCRIPTION
**Pull request by nvinuesa**
In order to start cleaning the legacy state out of Charm() calls, we need to first remove the one on the bootstrap worker so that Juju is able to deploy the controller charm without retrieving the charm from the legacy state.

This is possible because both for local and charmhub charms we have everything we need (namely URL and Meta), and therefore the only big change in this patch is to add the CharmURL to the attributes on AddApplication() and to move the URL() method out from CharmRef to CharmRefFull, this way we can pass everything during bootstrap.

## QA steps

Since the `AddApplication()` method has changed, with a new attribute, we must test bootstrapping and deploying (both local / ch charms):
```
$ juju bootstrap lxd c
$ juju download ubuntu
$ juju deploy ./ubuntu_r25.charm local
$ juju deploy ubuntu ch
```
Make sure there are no errors on the logs.

Also bootstrap with a local controller charm:
```
$ juju download juju-controller
Base "ubuntu@24.04" is not supported for charm "juju-controller", trying base "ubuntu@22.04"
Fetching charm "juju-controller" revision 91 using "stable" channel and base "amd64/ubuntu/22.04"
Install the "juju-controller" charm with:
    juju deploy ./juju-controller_r91.charm
$ juju bootstrap localhost c --build-agent --controller-charm-path ./juju-controller_r91.charm
```

Check if the controller is OK, simply with `juju status -m controller`.
## Links

**Jira card:** JUJU-7256

### CODE REVIEW

File: state/state.go [hunk 1]
```diff
@@ -908,12 +907,14 @@ type CharmRefFull interface {
 	Actions() *charm.Actions
 	Config() *charm.Config
 	Revision() int
+	URL() string
```
Comment (nvinuesa) [inline state/state.go:910]: This move allows us to call `AddApplication()` with a charm that isn't retrieved from state but instead is created from the archive, as we have during bootstrap.

**Review by SimonRichardson — COMMENTED** [review]
I'm happy with these changes, I would like to ensure that you can use a local `juju-controller` to bootstrap with (can you add that to the QA steps).

----

Others following along, the `apiserver/facades/controller/charmdownloader` is being removed soon. The `internal/charm/downloader` is also being superseded with a different one, that only downloads and doesn't attempt to store in the object store in one call.

**Review by jack-w-shaw — COMMENTED** [review]
LGTM. QA Passes

QA could do with bootstraping from a local controller charm. I have verified this myself

**Review by jack-w-shaw — APPROVED** [review]

**Comment by jack-w-shaw** [comment]
Getting a lot of unit test failures though!

File: state/state.go [hunk 2]
```diff
@@ -940,9 +941,6 @@ func (st *State) AddApplication(
 	if !names.IsValidApplication(args.Name) {
 		return nil, errors.Errorf("invalid name")
 	}
-	if args.Charm == nil {
-		return nil, errors.Errorf("charm is nil")
-	}
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply state/state.go:945]: It won't be used soon, as the next in the line is the removal of charm completely from the state package.

**Comment by nvinuesa** [comment]
@SimonRichardson I updated the QA steps to include local controller charm on bootstrap.

File: internal/bootstrap/deployer.go [hunk 3]
```diff
@@ -334,24 +334,20 @@ func (b *baseDeployer) DeployCharmhubCharm(ctx context.Context, arch string, bas
...
 
 	b.logger.Debugf("Successfully deployed charmhub Juju controller charm")
 
-	return curl.String(), &resOrigin, nil
+	return curl.String(), &resOrigin, ch, nil
 }
 
 // AddControllerApplication adds the controller application.
-func (b *baseDeployer) AddControllerApplication(ctx context.Context, curl string, origin corecharm.Origin, controllerAddress string) (Unit, error) {
-	ch, err := b.stateBackend.Charm(curl)
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply internal/bootstrap/deployer.go:351]: Done.

File: internal/bootstrap/deployer.go [hunk 4]
```diff
@@ -298,7 +298,7 @@ func (b *baseDeployer) DeployCharmhubCharm(ctx context.Context, arch string, bas
 		curl = charm.MustParseURL(b.controllerCharmName)
 	}
 	if err != nil {
-		return "", nil, errors.Trace(err)
+		return "", nil, nil, errors.Trace(err)
 	}
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply internal/bootstrap/deployer.go]: You are very much right, don't know why it was there on the first place. Removed.

**Review by SimonRichardson — APPROVED** [review]

**Pull request merged**